src/util/address.rs

```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+    V8,
+    /// Future (unsupported) version of Witness Program
+    V9,
+    /// Future (unsupported) version of Witness Program
+    V10,
+    /// Future (unsupported) version of Witness Program
+    V11,
+    /// Future (unsupported) version of Witness Program
+    V12,
+    /// Future (unsupported) version of Witness Program
+    V13,
+    /// Future (unsupported) version of Witness Program
+    V14,
+    /// Future (unsupported) version of Witness Program
+    V15,
+    /// Future (unsupported) version of Witness Program
+    V16,
+}
+
+impl Display for WitnessVersion {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", *self as u8)
+    }
+}
+
+
+impl FromStr for WitnessVersion {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let ver = u8::from_str(s).map_err(|_| Error::UnparsableWitnessVersion)?;
+        WitnessVersion::from_ver_u8(ver)
+    }
+}
+
+impl WitnessVersion {
+    /// Takes [`bech32::u5`] value and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_u5(value: ::bech32::u5) -> Result<Self, Error> {
+        WitnessVersion::from_ver_u8(value.to_u8())
+    }
+
+    /// Takes byte value and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_ver_u8(value: u8) -> Result<Self, Error> {
+        Ok(match value {
+            0 => WitnessVersion::V0,
+            1 => WitnessVersion::V1,
+            2 => WitnessVersion::V2,
+            3 => WitnessVersion::V3,
+            4 => WitnessVersion::V4,
+            5 => WitnessVersion::V5,
+            6 => WitnessVersion::V6,
+            7 => WitnessVersion::V7,
+            8 => WitnessVersion::V8,
+            9 => WitnessVersion::V9,
+            10 => WitnessVersion::V10,
+            11 => WitnessVersion::V11,
+            12 => WitnessVersion::V12,
+            13 => WitnessVersion::V13,
+            14 => WitnessVersion::V14,
+            15 => WitnessVersion::V15,
+            16 => WitnessVersion::V16,
+            wrong => Err(Error::InvalidWitnessVersion(wrong))?,
+        })
+    }
+
+    /// Takes bitcoin Script opcode and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_opcode(opcode: opcodes::All) -> Result<Self, Error> {
+        match opcode.into_u8() {
+            0 => Ok(WitnessVersion::V0),
+            ver if ver >= opcodes::all::OP_PUSHNUM_1.into_u8() && ver <= opcodes::all::OP_PUSHNUM_16.into_u8() =>
+                WitnessVersion::from_ver_u8(ver - opcodes::all::OP_PUSHNUM_1.into_u8() + 1),
+            _ => Err(Error::UnparsableWitnessVersion)
```

Comment:
Why not `InvalidWitnessVersion`?

Reply:
To my understanding `InvalidWitnessVersion` s for out of bounds numerical values and `UnparsableWitnessVersion` for things that aren't even valid numerical representations in this context (like all other op codes).

Reply:
Right, that makes sense.